server/src/main/java/org/elasticsearch/action/admin/indices/rollover/AutoShardCondition.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.rollover;
+
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.action.datastreams.autosharding.AutoShardingResult;
+import org.elasticsearch.action.datastreams.autosharding.AutoShardingType;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Condition for automatically increasing the number of shards for a data stream. The value is computed when the condition is
+ * evaluated.
+ */
+public class AutoShardCondition extends Condition<String> {
+    public static final String NAME = "auto_sharding";
+
+    public static final ParseField AUTO_SHARDING_TYPE = new ParseField("type");
+    public static final ParseField CURRENT_NUMBER_OF_SHARDS = new ParseField("current_number_of_shards");
+    public static final ParseField TARGET_NUMBER_OF_SHARDS = new ParseField("target_number_of_shards");
+    public static final ParseField WRITE_LOAD = new ParseField("write_load");
+    private final AutoShardingResult autoShardingResult;
+
+    public AutoShardCondition(AutoShardingResult autoShardingResult) {
+        super(NAME, Type.AUTOMATIC);
+        if (autoShardingResult.type().equals(AutoShardingType.INCREASE_SHARDS) == false
+            && autoShardingResult.type().equals(AutoShardingType.DECREASE_SHARDS) == false) {
+            throw new IllegalArgumentException("The autosharding condition only supports the INCREASE and DECREASE shards types");
+        }
+        this.value = toCSVRep(autoShardingResult);
+        this.autoShardingResult = autoShardingResult;
+    }
+
+    static String toCSVRep(AutoShardingResult autoShardingResult) {
+        return String.join(
+            ",",
+            List.of(
+                AUTO_SHARDING_TYPE.getPreferredName() + "=" + autoShardingResult.type(),
+                CURRENT_NUMBER_OF_SHARDS.getPreferredName() + "=" + autoShardingResult.currentNumberOfShards(),
+                TARGET_NUMBER_OF_SHARDS.getPreferredName() + "=" + autoShardingResult.targetNumberOfShards(),
+                WRITE_LOAD.getPreferredName() + "=" + autoShardingResult.writeLoad()
+            )
+        );
```

Comment:
You can use our `Strings` method for this with:
```suggestion
        return Strings.collectionToDelimitedString(
            Map.of(
                AUTO_SHARDING_TYPE.getPreferredName(), autoShardingResult.type(),
                CURRENT_NUMBER_OF_SHARDS.getPreferredName(), autoShardingResult.currentNumberOfShards(),
                TARGET_NUMBER_OF_SHARDS.getPreferredName(), autoShardingResult.targetNumberOfShards(),
                WRITE_LOAD.getPreferredName(), autoShardingResult.writeLoad()
            ).entrySet(),
            ","
        );
```

And avoid all the `"="` and `+`s